SUBMARINE-189. [WIP] Add proposal doc for submarine-server

jiwq opened the pull request:
### What is this PR for?
Add submarine server architecture doc.
https://docs.google.com/document/d/1vCjt2z3DoagfeaVQV0nfyWqSYLymoP53rDcPcz-YcSc/edit?usp=sharing

### What type of PR is it?
[Documentation]

### Todos
See TODO label

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-189

### How should this be tested?
https://travis-ci.org/jiwq/hadoop-submarine/builds/601922439

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

tangzhankun commented on docs/design/submarine-server/architecture.md:
```diff
@@ -0,0 +1,126 @@
+  under the License.
+  -->
+## Motivation
+Due to the single responsibility of considering different components, and nowadays the core module depends the hadoop libs which can be setted optional.
```
This is a little confusing to me. Could you please rephrase it? The motivation in my mind for the submarine server is to hide the underlying YARN/k8s so that submarine-server can act as a job server and manage all jobs submitted.

jiwq replied:
Yes, I forgot to update it. I will update it later.

tangzhankun commented on docs/design/submarine-server/architecture.md:
````diff
@@ -0,0 +1,126 @@
+ +-----------+                                                  |     K8s Cluster     |
+    client                          server                      +---------------------+
+ ```
+We propose to split the core module to two modules, CLI and server as shown in FIG. The client call the REST API to submit and retrieve the job info. 
````
"the core module" -> "the original core module in the old layout"

jiwq replied:
Clever description, I will modify the doc.

tangzhankun commented on docs/design/submarine-server/architecture.md:
```diff
@@ -0,0 +1,126 @@
+The server-sdk will be added in the future. It is the official client for Submarine Server API.
+
+#### REST
+It provides the RESTful API for users, consists of submit job and get job info features and so on. Often used by CLI and workbench etc. The submarine client project should implement all APIs.
```
Can we define the REST APIs here? When defining it, we should consider the current YAML files format and the tf-operator job spec, pyTorch operator job spec and also horovod spec. Please draft it if you've extra cycles and let's review it together.

jiwq replied:
As the Birds Eyes View Architecture, I think we should keep it simple and should clever describe the architecture.  About the each component's detailed design maybe we can propose inside alone doc in the same location.

tangzhankun commented on docs/design/submarine-server/architecture.md:
```diff
@@ -0,0 +1,126 @@
+#### Monitor
+We need a monitor to track the job life cycle and record the main events and key info in running. If the client register to server, the monitor will called the pusher to push the info to client.
+
+#### Translator
```
I'm wondering if this translator logic should be inside each submitter. Thoughts?

jiwq replied:
We had only one translator as the main component of submarine-server. It translates the submarine job spec to the corresponding cluster's job spec.  Data flow as below:
```
                                         +----------------+
            +------------+ tf-spec-yarn  |                |
            |            +-------------->+ submitter-yarn |
            |            |               |                |
    spec    |            |               +----------------+
+---------->+ translator |
            |            |               +----------------+
            |            |  tf-spec-k8s  |                |
            |            +-------------->+ submitter-k8s  |
            +------------+               |                |
                                         +----------------+
```
Any thoughts? If no other doubts, I'll update the doc.

tangzhankun replied:
what if there's another k8s runtime implemented a new CRD? I'm not clear of the core responsibility of submitter-k8s. Is it a k8s client wrapper?

tangzhankun commented on docs/design/submarine-server/architecture.md:
```diff
@@ -0,0 +1,126 @@
+#### Pusher(TODO)
+Pusher provides the API to create collaboration & communication feature. It consists of server and SDK modules. The SDK used by client to keep the connect to server and receive info.
+
+#### DB (optional)
```
The DB can be used to do submarine-server failure recovery.

jiwq replied:
Yes, excepts that it can store the job history.

tangzhankun commented:
Could you paste the link of the google doc in the description so that we can edit or comment directly? @jiwq

tangzhankun commented:
I drafted a google doc here to help design the generic job spec and REST API easily. Please share your thoughts. @jiwq @liuxunorg @yuanzac
https://docs.google.com/document/d/1kd-5UzsHft6gV7EuZiPXeWIKJtPqVwkNlqMvy0P_pAw/edit?usp=sharing

jiwq commented:
@tangzhankun Thanks for your reviewing. I had paste the link in the description and addressed all the comments. About the design of the generic job spec and REST API, I'll comment in the google doc.

yuanzac commented on docs/design/submarine-server/architecture.md:
```diff
@@ -0,0 +1,75 @@
+# Submarine Server
+
+## Motivation
+Nowadays we use the client to submit the machine learning job and the cluster config is stored at the client. It’s not secure and difficult to use. So we should hide the underlying cluster system such as YARN/K8s.
```
It is better to change "not secure" to "insecure".

jiwq replied:
DONE

yuanzac commented on docs/design/submarine-server/architecture.md:
```diff
@@ -0,0 +1,75 @@
+
+The JobManager has two main components: auxiliary manager and job monitor.
+
+#### AuxMgr
```
To void dependency conflicts, is AuxMgr is mainly used for managing different classloaders of submitter plugins? It's ok for me, but I'm not sure about the name. @liuxunorg, @tangzhankun can you help to take a look at this part?

liuxunorg replied:
I think `auxiliary ` is a very vague term. maybe the `pluginManger` will be clearer?

jiwq replied:
Agree to adopt the `PluginManager` term which abbr is `PlugMgr`.

yuanzac commented:
Thanks @jiwq for the contribution and @tangzhankun, @liuxunorg for the review.

yuanzac commented:
Thanks @jiwq for the contribution and @tangzhankun, @liuxunorg for the review.
